dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/h12/grpc/GrpcCompositeCodec.java

```diff
@@ -16,123 +16,103 @@
 
 public class GrpcCompositeCodec implements HttpMessageCodec {
 
-    private final ProtobufHttpMessageCodec protobufHttpMessageCodec;
+    private static final String PACKABLE_METHOD_CACHE = "PACKABLE_METHOD_CACHE";
 
-    private final WrapperHttpMessageCodec wrapperHttpMessageCodec;
+    private final URL url;
 
-    public GrpcCompositeCodec(
-            ProtobufHttpMessageCodec protobufHttpMessageCodec, WrapperHttpMessageCodec wrapperHttpMessageCodec) {
-        this.protobufHttpMessageCodec = protobufHttpMessageCodec;
-        this.wrapperHttpMessageCodec = wrapperHttpMessageCodec;
-    }
+    private final FrameworkModel frameworkModel;
+
+    private final String mediaType;
 
-    public void setEncodeTypes(Class<?>[] encodeTypes) {
-        this.wrapperHttpMessageCodec.setEncodeTypes(encodeTypes);
+    private PackableMethod packableMethod;
+
+    public GrpcCompositeCodec(URL url, FrameworkModel frameworkModel, String mediaType) {
+        this.url = url;
+        this.frameworkModel = frameworkModel;
+        this.mediaType = mediaType;
     }
 
-    public void setDecodeTypes(Class<?>[] decodeTypes) {
-        this.wrapperHttpMessageCodec.setDecodeTypes(decodeTypes);
+    public void loadPackableMethod(MethodDescriptor methodDescriptor) {
+        if (methodDescriptor instanceof PackableMethod) {
+            packableMethod = (PackableMethod) methodDescriptor;
+            return;
+        }
+        Map<MethodDescriptor, PackableMethod> cacheMap = (Map<MethodDescriptor, PackableMethod>) url.getServiceModel()
+                .getServiceMetadata()
+                .getAttributeMap()
+                .computeIfAbsent(PACKABLE_METHOD_CACHE, k -> new ConcurrentHashMap<>());
+        packableMethod = cacheMap.computeIfAbsent(methodDescriptor, md -> frameworkModel
+                .getExtensionLoader(PackableMethodFactory.class)
+                .getExtension(ConfigurationUtils.getGlobalConfiguration(url.getApplicationModel())
+                        .getString(DUBBO_PACKABLE_METHOD_FACTORY, DEFAULT_KEY))
+                .create(methodDescriptor, url, mediaType));
     }
 
     @Override
     public void encode(OutputStream outputStream, Object data, Charset charset) throws EncodeException {
-        // protobuf
-        // TODO int compressed = Identity.MESSAGE_ENCODING.equals(requestMetadata.compressor.getMessageEncoding()) ? 0 :
-        // 1;
         try {
-            int compressed = 0;
-            outputStream.write(compressed);
-            if (isProtobuf(data)) {
-                ProtobufWriter.write(protobufHttpMessageCodec, outputStream, data);
-                return;
-            }
-            // wrapper
-            wrapperHttpMessageCodec.encode(outputStream, data);
-        } catch (IOException e) {
+            outputStream.write(0);
+            byte[] bytes = packableMethod.packResponse(data);
+            writeLength(outputStream, bytes.length);
+            outputStream.write(bytes);
```

Comment:
**Scalability Issue**: Refactoring to use a more modular and extensible codec system, such as replacing specific codec implementations with a more generic, pluggable mechanism, can significantly enhance the scalability of the system. This allows for easier addition of new codecs and serialization mechanisms without modifying the core logic, which is crucial for maintaining a scalable, extensible system. <br> **Fix**: Implement a codec registry or factory pattern that allows for runtime registration and retrieval of codecs. This pattern should support lazy loading and initialization of codecs to minimize startup time and memory footprint, enhancing the scalability of the system. <br> **Code Suggestion**: 
 ```
 public class GrpcCompositeCodec implements HttpMessageCodec {

     private final URL url;

     private final FrameworkModel frameworkModel;

     private final String mediaType;

     private PackableMethod packableMethod;

     private static final Map<MethodDescriptor, PackableMethod> cacheMap = new ConcurrentHashMap<>();

     public GrpcCompositeCodec(URL url, FrameworkModel frameworkModel, String mediaType) {
         this.url = url;
         this.frameworkModel = frameworkModel;
         this.mediaType = mediaType;
     }

     public void loadPackableMethod(MethodDescriptor methodDescriptor) {
         packableMethod = cacheMap.computeIfAbsent(methodDescriptor, md -> frameworkModel
                 .getExtensionLoader(PackableMethodFactory.class)
                 .getExtension(ConfigurationUtils.getGlobalConfiguration(url.getApplicationModel())
                         .getString(DUBBO_PACKABLE_METHOD_FACTORY, DEFAULT_KEY))
                 .create(methodDescriptor, url, mediaType));
     }

     @Override
     public void encode(OutputStream outputStream, Object data, Charset charset) throws EncodeException {
         ...
     }

     @Override
     public Object decode(InputStream inputStream, Class<?> targetType, Charset charset) throws DecodeException {
         ...
     }

     ...
 }
 ```